### PR TITLE
[2109] [7538719,7834158,7834080] Broder group controls improvements

### PR DESCRIPTION
**Pull request by v-soukum**
## Updated following controls

1. ADO_AgentPool_AuthZ_Restrict_Broader_Group_Access
2. ADO_Build_AuthZ_Restrict_Broader_Group_Access
3. ADO_Feed_AuthZ_Restrict_Broader_Group_Access
4. ADO_SecureFile_AuthZ_Restrict_Broader_Group_Access
5. ADO_Environment_AuthZ_Restrict_Broader_Group_Access
6. ADO_Release_AuthZ_Restrict_Broader_Group_Access
7. ADO_ServiceConnection_AuthZ_Restrict_Broader_Group_Access
8. ADO_VariableGroup_AuthZ_Restrict_Broader_Group_Access
9. ADO_VariableGroup_AuthZ_Restrict_Broader_Group_Access_On_VG_With_Secrets
10. ADO_Project_AuthZ_Restrict_Broader_Group_Access_on_Agentpool
11. ADO_Project_AuthZ_Restrict_Broader_Group_Access_on_Builds
12. ADO_Project_AuthZ_Restrict_Broader_Group_Access_on_SecureFile
13. ADO_Project_AuthZ_Restrict_Broader_Group_Access_on_Env
14. ADO_Project_AuthZ_Restrict_Broader_Group_Access_on_Releases
15. ADO_Project_AuthZ_Restrict_Broader_Group_Access_on_SvcConn
16. ADO_Project_AuthZ_Restrict_Broader_Group_Access_on_VarGrp

## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.AgentPool.ps1 [hunk 1]
```diff
@@ -410,52 +410,54 @@ class AgentPool: ADOSVTBase
...
             $controlResult.VerificationResult = [VerificationResult]::Failed
+            $restrictedBroaderGroups = @{}
+            $restrictedBroaderGroupsForAgentPool = $this.ControlSettings.AgentPool.RestrictedBroaderGroupsForAgentPool;
+            $restrictedBroaderGroupsForAgentPool.psobject.properties | foreach { $restrictedBroaderGroups[$_.Name] = $_.Value }
+            if (($this.AgentObj.Count -gt 0) -and [Helpers]::CheckMember($this.AgentObj, "identity")) {
+                # match all the identities added on agentpool with defined restricted list
+                $roleAssignmentsToCheck = $this.AgentObj
+                $restrictedGroups = @()
+                if ($this.checkInheritedPermissionsPerAgentPool -eq $false) {
+                    $roleAssignmentsToCheck = $this.AgentObj | where-object { $_.access -ne "inherited" }
```
Comment (v-himkam) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.AgentPool.ps1:421]: @()

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.AgentPool.ps1 [hunk 2]
```diff
@@ -410,52 +410,54 @@ class AgentPool: ADOSVTBase
...
+                    $broaderGroupsWithExcessiveMembers = @([ControlHelper]::FilterBroadGroupMembers($restrictedGroups, $true))
+                    $restrictedGroups = @($restrictedGroups | Where-Object {$broaderGroupsWithExcessiveMembers -contains $_.Name})
+                }
+                $restrictedGroupsCount = $restrictedGroups.Count
+                # fail the control if restricted group found on agentpool
+                if ($restrictedGroupsCount -gt 0) {
+                    $controlResult.AddMessage([VerificationResult]::Failed, "Count of broader groups that have excessive permissions on agent pool: $($restrictedGroupsCount)");
+                    $formattedGroupsData = $restrictedGroups | Select @{l = 'Group'; e = { $_.Name} }, @{l = 'Role'; e = { $_.Role } }
+                    $backupDataObject = $restrictedGroups | Select @{l = 'Group'; e = { $_.Name} },@{l = 'Id'; e = { $_.Id } }, @{l = 'Role'; e = { $_.Role } }
+                    $formattedGroupsTable = ($formattedGroupsData | Out-String)
```
Comment (v-himkam) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.AgentPool.ps1:438]: why FT -AutoSize is not used for display ?

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.AgentPool.ps1 [hunk 3]
```diff
@@ -410,52 +410,54 @@ class AgentPool: ADOSVTBase
...
                 }
-                $controlResult.AddMessage("`nNote:`nThe following groups are considered 'broad' which should not have user/administrator privileges: `n$($restrictedBroaderGroupsForAgentPool | FT | out-string )`n");
             }
             else {
-                $controlResult.AddMessage([VerificationResult]::Error, "List of restricted broader groups for agent pool is not defined in control settings for your organization.");
+                $controlResult.AddMessage([VerificationResult]::Passed, "No groups have given access to agent pool.");
+                $controlResult.AdditionalInfoInCSV = "NA";
             }
+            $displayObj = $restrictedBroaderGroups.Keys | Select-Object @{Name = "Broader Group"; Expression = {$_}}, @{Name = "Excessive Permissions"; Expression = {$restrictedBroaderGroups[$_] -join ', '}}
+            $controlResult.AddMessage("Note:`nThe following groups are considered 'broad' which should not excessive permissions: `n$($displayObj | FT | out-string -width 512)");
```
Comment (v-himkam) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.AgentPool.ps1:460]: -Autosize

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Build.ps1 [hunk 4]
```diff
@@ -1500,8 +1502,16 @@ class Build: ADOSVTBase
...
                             }
+
+                            if ($this.ControlSettings.CheckForBroadGroupMemberCount -and $filteredBroaderGroupList.Count -gt 0)
+                            {
+                                $broaderGroupsWithExcessiveMembers = @([ControlHelper]::FilterBroadGroupMembers($filteredBroaderGroupList, $false))
+                                $groupsWithExcessivePermissionsList = @($groupsWithExcessivePermissionsList | Where-Object {$broaderGroupsWithExcessiveMembers -contains $_.Group})
+                            }
+
                             if ($groupsWithExcessivePermissionsList.count -gt 0) {
                                 #TODO: Do we need to put state object?
```
Comment (v-himkam) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.Build.ps1:1516]: if this #TODO done ?

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1 [hunk 5]
```diff
@@ -297,7 +305,8 @@ class CommonSVTControls: ADOSVTBase {
                     $controlResult.AddMessage([VerificationResult]::Passed,  "Feed is not granted with administrator/contributor/collaborator permission to broad groups.");
                     $controlResult.AdditionalInfoInCSV = "NA";
                 }
-                $controlResult.AddMessage("`nNote: `nThe following groups are considered 'broader groups': `n$($restrictedBroaderGroupsForFeeds | FT | out-string)");
+                $displayObj = $restrictedBroaderGroups.Keys | Select-Object @{Name = "Broader Group"; Expression = {$_}}, @{Name = "Excessive Permissions"; Expression = {$restrictedBroaderGroups[$_] -join ', '}}
+                $controlResult.AddMessage("`nNote: `nThe following groups are considered 'broader groups': `n$($displayObj | FT | out-string)");
```
Comment (v-himkam) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1:309]: -Autosize

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1 [hunk 6]
```diff
@@ -560,22 +579,30 @@ class CommonSVTControls: ADOSVTBase {
...
+                    $display = ($environmentWithBroaderGroup |  FT Name, Role -AutoSize | Out-String -Width 512)
                     $controlResult.AddMessage("`nList of groups: ", $display)
                 }
                 else
                 {
                     $controlResult.AddMessage([VerificationResult]::Passed,  "Environment is not granted with user/administrator permission to broad groups.");
                 }
-                $controlResult.AddMessage("`nNote: `nThe following groups are considered 'broader groups': `n$($restrictedBroaderGroupsForEnvironment | FT | out-string)");
+                $displayObj = $restrictedBroaderGroups.Keys | Select-Object @{Name = "Broader Group"; Expression = {$_}}, @{Name = "Excessive Permissions"; Expression = {$restrictedBroaderGroups[$_] -join ', '}}
+                $controlResult.AddMessage("`nNote: `nThe following groups are considered 'broader groups': `n$($displayObj | FT | out-string)");
```
Comment (v-himkam) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1:605]: -Autosize

**Review by arvindsingh-cloud — DISMISSED** [review]

**Review by Garima-msft — APPROVED** [review]

**Review by arvindsingh-cloud — APPROVED** [review]

**Pull request merged**